vm/systemSmartContracts/stakingWaitingList.go

```diff
@@ -0,0 +1,1170 @@
+package systemSmartContracts
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"math"
+	"math/big"
+	"strconv"
+
+	"github.com/ElrondNetwork/elrond-go/common"
+	"github.com/ElrondNetwork/elrond-go/vm"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+const waitingListHeadKey = "waitingList"
+const waitingElementPrefix = "w_"
+
+type waitingListReturnData struct {
+	blsKeys         [][]byte
+	stakedDataList  []*StakedDataV2_0
+	lastKey         []byte
+	afterLastJailed bool
+}
+
+func (s *stakingSC) processStake(blsKey []byte, registrationData *StakedDataV2_0, addFirst bool) error {
+	if registrationData.Staked {
+		return nil
+	}
+
+	registrationData.RegisterNonce = s.eei.BlockChainHook().CurrentNonce()
+	if !s.canStake() {
+		s.eei.AddReturnMessage(fmt.Sprintf("staking is full key put into waiting list %s", hex.EncodeToString(blsKey)))
+		err := s.addToWaitingList(blsKey, addFirst)
+		if err != nil {
+			s.eei.AddReturnMessage("error while adding to waiting")
+			return err
+		}
+		registrationData.Waiting = true
+		s.eei.Finish([]byte{waiting})
+		return nil
+	}
+
+	if !s.flagStakingV4.IsSet() {
+		err := s.removeFromWaitingList(blsKey)
+		if err != nil {
+			s.eei.AddReturnMessage("error while removing from waiting")
+			return err
+		}
+	}
+	s.addToStakedNodes(1)
+	s.activeStakingFor(registrationData)
+
+	return nil
+}
+
+func (s *stakingSC) unStakeAtEndOfEpoch(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```

Comment:
this again should be in the base file - this is not legacy - this will work with stakingV4 as well.

Reply:
Right, when I started this refactor, I just had in mind "files with/without waiting list", not legacy. Will move it back